app/com/linkedin/drelephant/analysis/AnalyticJobGeneratorHadoop2.java

```diff
@@ -122,40 +124,60 @@ public void configure(Configuration configuration)
    * @throws AuthenticationException
    */
   @Override
-  public List<AnalyticJob> fetchAnalyticJobs()
+  public List<AnalyticJob> fetchCompletedAnalyticJobs(long from, long to)
       throws IOException, AuthenticationException {
     List<AnalyticJob> appList = new ArrayList<AnalyticJob>();
 
-    // There is a lag of job data from AM/NM to JobHistoryServer HDFS, we shouldn't use the current time, since there
-    // might be new jobs arriving after we fetch jobs. We provide one minute delay to address this lag.
-    _currentTime = System.currentTimeMillis() - FETCH_DELAY;
-    updateAuthToken();
-
-    logger.info("Fetching recent finished application runs between last time: " + (_lastTime + 1)
-        + ", and current time: " + _currentTime);
-
     // Fetch all succeeded apps
     URL succeededAppsURL = new URL(new URL("http://" + _resourceManagerAddress), String.format(
             "/ws/v1/cluster/apps?finalStatus=SUCCEEDED&finishedTimeBegin=%s&finishedTimeEnd=%s",
-            String.valueOf(_lastTime + 1), String.valueOf(_currentTime)));
-    logger.info("The succeeded apps URL is " + succeededAppsURL);
-    List<AnalyticJob> succeededApps = readApps(succeededAppsURL);
+            String.valueOf(from), String.valueOf(to)));
+    //logger.info("The succeeded apps URL is " + succeededAppsURL);
+    JsonNode successList = readApps(succeededAppsURL).path("apps").path("app");
+    List<AnalyticJob> succeededApps = filterDuplicates(successList);
     appList.addAll(succeededApps);
 
     // Fetch all failed apps
     URL failedAppsURL = new URL(new URL("http://" + _resourceManagerAddress), String.format(
             "/ws/v1/cluster/apps?finalStatus=FAILED&finishedTimeBegin=%s&finishedTimeEnd=%s",
-            String.valueOf(_lastTime + 1), String.valueOf(_currentTime)));
-    List<AnalyticJob> failedApps = readApps(failedAppsURL);
-    logger.info("The failed apps URL is " + failedAppsURL);
+            String.valueOf(from), String.valueOf(to)));
+    //logger.info("The failed apps URL is " + failedAppsURL);
+    JsonNode failList = readApps(failedAppsURL).path("apps").path("app");
+    List<AnalyticJob> failedApps = filterDuplicates(failList);
     appList.addAll(failedApps);
 
     // Append promises from the retry queue at the end of the list
     while (!_retryQueue.isEmpty()) {
       appList.add(_retryQueue.poll());
     }
 
-    _lastTime = _currentTime;
+    logger.info("Fetched " + appList.size() + " completed applications.");
```

Comment:
I think there is value in separating the succeeded and failed applications list size